feat(lncli): Add --route_hints flag to sendpayment and queryroutes

appilon opened the pull request:
## Change Description

This PR implements the `--route_hints` flag for the `lncli sendpayment` and `lncli queryroutes` commands, addressing issue #6601. This allows users to manually provide route hints, typically obtained from invoices for private channels.

I did not see an existing integration test that covered route_hints so I added one that invoked SendPayment over RPC, while it does not directly test the argparsing added in this PR, I found it a valuable add to the testsuite.

As this is my first contribution to this project and bitcoin development in general, I faced substantial payment failed issues in testing via `itest`, it turned out to be [this issue](https://github.com/lightningnetwork/lnd/issues/9336). Thinking there was some kind of routing issue, I did make a change to `routing/router.go` to include more data in the finalHop struct, feel free to request I remove that commit if its incorrect or just too risky in general so deep in the callstack.

Closes #6601

## Steps to Test

1.  **Verify `lncli` argument parsing:**
    *   Run `lncli --network=simnet --no-macaroons sendpayment --dest=<dummy_pubkey> --amt=1 --payment_hash=<dummy_hash> --route_hints='[{"hop_hints":[{"node_id":"03aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "chan_id":1, "fee_base_msat": 1, "fee_proportional_millionths": 1, "cltv_expiry_delta": 1}]}]'` (using dummy values). Expect a connection error (e.g., `connection refused`), indicating parsing succeeded.
    *   Run the same command but with invalid JSON in `--route_hints` (e.g., remove a quote). Expect a JSON parsing error.
    *   Repeat similar basic parsing checks for `lncli queryroutes --route_hints='...'`.

2.  **Run the new integration test:**
    *   `make itest ITEST_FLAGS='-test.run=TestLightningNetworkDaemon/tranche00/43-of-279/btcd/sendpayment_with_route_hints'` (not sure if ordering is consistent for all users but I was able to rerun test over and over with that command)

     ```--- PASS: TestLightningNetworkDaemon/tranche00/43-of-279/btcd/sendpayment_with_route_hints (21.63s)```

## Pull Request Checklist

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included. *(CLI parsing tests implicit in Steps to Test, positive path covered by itest)*
- [ ] Bug fixes contain tests triggering the bug to prevent regressions. *(FindRoute fix implicitly tested by the new itest)*

### Code Style and Documentation
- [x] The change is not [insubstantial](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#substantial-contributions-only).
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/development_guidelines.md#code-documentation-and-commenting) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/development_guidelines.md#ideal-git-commit-structure).
- [x] Any new logging statements use an appropriate subsystem and logging level. *(No new logging added)*
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file. *(N/A - flags added to existing commands)*
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes. *(Release notes entry needed)*

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

appilon commented:
@saubyk I fixed the linting errors I believe (strange though `cmd_payments.go` does not have strict 80 char limit everywhere? I don't mind following the rule of course). As for the unit test failures, I don't see them locally? I did have a commit earlier that caused some failures but I dropped it, perhaps the unit tests didn't re-run or are flakey.

appilon commented:
I have a question for the reviewer, I added the argparsing to anywhere where SendPayment is called, however as a lightning newbie I'm unsure how that would interact with `--pay_req/payinvoice`? Would it interfere with route hints included in the payment req string (does payment req string contain route hints? Gemini says yes, but I'm learning as I go). It might be safer to remove the argument support for `payinvoice/sendpayment --pay-req`

appilon commented:
@saubyk PR should be ready for review and re-running of CI.

appilon commented:
Sorry about lint error 🤦 forgot to re-rerun locally. fixed now

saubyk commented:
> @saubyk PR should be ready for review and re-running of CI.

Hi @appilon we are currently busy with wrapping up release 0.19. Should be able to get someone to look at the pr in the coming weeks. Meanwhile, anyone else who has bandwidth & interest should feel free to pick this one up

Lokeshranjan8 commented on itest/lnd_payment_test.go:
```diff
@@ -1396,3 +1397,214 @@ func testSendPaymentKeysendMPPFail(ht *lntest.HarnessTest) {
+		},
+		RouteHints:     hints,
+		FeeLimitSat:    int64(chanAmt),
+		TimeoutSeconds: 60,
```
Here we're using a 60s timeout. Since this test is running on a local harness with predictable topology and no external delays, would reducing this to something like 20–30s still keep things safe while improving test turnaround time? Could help with CI speedups

Lokeshranjan8 commented on itest/lnd_payment_test.go:
```diff
@@ -1396,3 +1397,214 @@ func testSendPaymentKeysendMPPFail(ht *lntest.HarnessTest) {
+	// Setup a three-node network: Alice -> Bob -> Carol.
+	alice := ht.NewNode("Alice", nil)
+	defer ht.Shutdown(alice)
+	bob := ht.NewNode("Bob", nil)
+	defer ht.Shutdown(bob)
+	// Ensure Carol accepts keysend payments.
+	carol := ht.NewNode("Carol", []string{"--accept-keysend"})
+	defer ht.Shutdown(carol)
+
+	ht.ConnectNodes(alice, bob)
+	ht.ConnectNodes(bob, carol)
+
+	// Fund Alice and Bob.
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, alice)
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, bob)
+
+	// Open channels: Alice -> Bob (public), Bob -> Carol (private).
+	const chanAmt = btcutil.Amount(1_000_000)
+	aliceBobChanPoint := ht.OpenChannel(
+		alice, bob, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(alice, aliceBobChanPoint)
+	bobCarolChanPoint := ht.OpenChannel(bob, carol,
+		lntest.OpenChannelParams{Amt: chanAmt, Private: true},
+	)
+	defer ht.CloseChannel(bob, bobCarolChanPoint)
+
+	// Manually create route hints for the private Bob -> Carol channel.
+	bobChan := ht.GetChannelByChanPoint(bob, bobCarolChanPoint)
+	require.NotNil(ht, bobChan, "Bob should have channel with Carol")
+
+	hints := []*lnrpc.RouteHint{{HopHints: []*lnrpc.HopHint{
+		{
+			NodeId:                    bob.PubKeyStr,
+			ChanId:                    bobChan.ChanId,
+			FeeBaseMsat:               1000,
+			FeeProportionalMillionths: 1,
+			CltvExpiryDelta:           routing.MinCLTVDelta,
+		}}},
+	}
```
one more thing ,I noticed that the setup code for the three-node network [Alice, Bob, Carol] and route hints is repeated across multiple test functions. Since the carol node setup for keysend payments is unique to {testSendPaymentRouteHintsKeysend}, I suggest refactoring the common setup (Alice and Bob node creation, funding, and channel opening) into a helper function.Let me know your thoughts on this approach :)

appilon replied:
I just checked the test without the accept-keysend flag and it still works, so I can definitely refactor the setup code

MPins reviewed (commented):
Good Job :+1: 

I’ve successfully run the integration tests and also performed some manual tests using a regtest environment — including `sendpayment` with `keysend` and `queryroutes`.

The only thing missing now are the manual AMP tests, which I’ll be running in the next few days. Everything seems to be working properly so far, but I noticed a logic issue related to when a `pay_req` is provided — I believe it should be handled a bit differently.

appilon commented on cmd/commands/cmd_payments.go:
```diff
@@ -473,6 +483,31 @@ func SendPayment(ctx *cli.Context) error {
 	req.PaymentAddr = payAddr
 
+	if ctx.IsSet("route_hints") {
+		// Route hints can only be used with a manual keysend or amp
+		// payment. It should probably not be allowed with a payment
+		// request.
+		if ctx.IsSet("pay_req") ||
+			(!ctx.Bool("keysend") && !ctx.Bool(ampFlag.Name)) {
```
I was aware of this and mentioned it in my PR description, I figured this would be the feedback. My only reasoning for having the check was if there was a refactor, the entire `IsSet("route_hints")` block could be moved around wholesale and not be concerned with things happening upstream, but I'm totally fine removing the check 👍

appilon commented on itest/lnd_payment_test.go:
```diff
@@ -1396,3 +1397,214 @@ func testSendPaymentKeysendMPPFail(ht *lntest.HarnessTest) {
+	// Send AMP payment and assert success
+	ht.AssertPaymentStatusFromStream(
+		alice.RPC.SendPayment(sendReq),
+		lnrpc.Payment_SUCCEEDED,
```
Can do for both requests, I assume based on precedant I've seen there would be another separate itest with the same naming just `"Fail"` appended?

appilon replied:
Actually if acceptable by reviewers, it's probably easier just to run the request without the hints, assert it fails (its gunna hit the timeout, I've now set that to 30sec) and then run the request again with the hints and assert it passes (and avoid all the repeated setup code).

appilon commented on itest/lnd_payment_test.go:
```diff
@@ -1396,3 +1397,214 @@ func testSendPaymentKeysendMPPFail(ht *lntest.HarnessTest) {
+	hop2 := routes.Routes[0].Hops[1]
+
+	require.Equal(ht, bob.PubKeyStr, hop1.PubKey, "Hop 1 should be Bob")
+	require.Equal(ht, carol.PubKeyStr, hop2.PubKey, "Hop 2 should be Carol")
```
Can do

appilon commented:
@MPins @Lokeshranjan8 All review comments have been addressed, thank you for your time reviewing.

MPins commented:
> @MPins @Lokeshranjan8 All review comments have been addressed, thank you for your time reviewing.

Hello @Lokeshranjan8,

Looking deeper into the --amp case, I believe it should be treated like a regular invoice, where route hints are included at creation time, as defined in [BOLT #11](https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#tagged-fields).

Therefore, we should keep this change limited to `sendpayment --keysend` and `queryroutes`.

It would be great to get @saubyk 's input on this as well.

appilon commented:
@MPins I'm a lot less familiar with the specs and protocols so if AMP payments already handle route hinting similar to a bolt 11 pay req then yes I agree I can limit the flag to just support `--keysend`, I'd love a core contributor to weigh in @saubyk . Thank you everyone. I'll prep a release note in the meantime until we wait for the final decision.

MPins commented:
> ase note in the meantime until we wait for the final decision.

Hello, I just ran a test with AMP invoices, and they do accept route hints when creating the invoice.
Therefore, I think you should move forward and limit this change to sendpayment --keysend and queryroutes only.

Also, I recommend reading the [ideal git commit structure guide](https://github.com/lightningnetwork/lnd/blob/master/docs/development_guidelines.md#ideal-git-commit-structure) to improve the commit style.

MPins reviewed (commented):
Hello, I just ran a test with AMP invoices, and they do accept route hints when creating the invoice.
Therefore, I think you should move forward and limit this change to sendpayment --keysend and queryroutes only.

Also, I recommend reading the [ideal git commit structure guide](https://github.com/lightningnetwork/lnd/blob/master/docs/development_guidelines.md#ideal-git-commit-structure) to improve the commit style.

appilon commented:
@MPins Addressed, as for the commit, is style in general (like tense or how I broke up the sentences/paragraphs) an issue or the fact that it's one commit? I know the header is > 50 chars, I kept it under 72, If 50 chars is some hard mandate I'll try my best to compress it.

MPins commented on cmd/commands/cmd_payments.go:
```diff
@@ -254,6 +255,15 @@ var SendPaymentCommand = cli.Command{
 		},
+		cli.StringFlag{
+			Name: "route_hints",
+			Usage: "route hints for reaching the destination of a" +
+				" manual --keysend payment. eg: " +
+				`[{"hop_hints":[{"node_id":"A","chan_id":1,` +
+				`"fee_base_msat":2,` +
+				`"fee_proportional_millionths":3,` +
+				`"cltv_expiry_delta":4}]}]`,
```
Consider avoiding starting the line with a space, mixing " and ` in the same string concatenation, and preserving the single quotes (') around the example shown to the user.

MPins reviewed (commented):
Consider splitting this commit into three commits:
– the functional change
– the test coverage
– the documentation update

lightninglabs-deploy commented:
@lokeshranjan8: review reminder
@mpins: review reminder

MPins reviewed (approved):
I have just one final comment about the commit message. LGTM :tada: 

@saubyk the CI needs approval to run.

appilon commented:
@ziggie1984 I don't see why that itest failed in the postgres environment, it seems unrelated to my changes. I'm assuming it's simply a flakey test, I am trying to re-run the fullsuite with the same parameters locally.

ziggie1984 commented on cmd/commands/cmd_payments.go:
```diff
@@ -473,6 +483,26 @@ func SendPayment(ctx *cli.Context) error {
 
 	req.PaymentAddr = payAddr
 
+	if ctx.IsSet("route_hints") {
```
We should allow route_hints always if NO payment_request was provided, because we also wanna use it via AMP Sends not only keysends.

appilon replied:
there's an exit early upstream if pay_req is set, I left the logic to error if neither are set (a more explicit code pathway). Hope that's okay.

appilon replied:
Route hinting did not appear to work with standard payments in my testing, hence why I restricted it to AMP/keysend. @MPins felt that it should not be added to AMP to avoid conflicts for when they are embedded in the AMP request (similar to pay_req). I'm new to lightning dev so I'm not familiar with all the nuances of the protocol.

ziggie1984 replied:
So in general we should only prevent it if there is a reason for it, right now how you implemented it, it would accept AMP spontaneous payments, normal AMP payments (without a payment request) and keysend non-AMP payments. So the only thing which would be missing are legacy payments or MPP payments which specify a payment hash and destination manually (tho they would likely fail on the receiver side if a corresponding invoice for this hash) can be found. So in any way I think we can just add the route_hints here because the later cash is very unlikely to happen and should not be a problem for LND to handle it properly.

Would be interesting which problems you had enabling it for everything.

appilon replied:
I used the itest harness to recreate the various payment scenarios, its been a while now I can't remember all the various things I tried but it would just hang indefinitely and eventually fail with no route found for standard payments (normal and legacy). It passed for both keysend and amp which is why I had enabled it for those cases.

My guess is it's a rather safe thing to have enabled across all payment types and if there does happen to be a scenario where it's not working, someone can file an issue. I will just remove any restriction.

ziggie1984 commented on cmd/commands/cmd_payments.go:
```diff
@@ -1248,6 +1287,19 @@ func queryRoutes(ctx *cli.Context) error {
 		BlindedPaymentPaths: blindedRoutes,
 	}
 
+	if ctx.IsSet("route_hints") {
```
route hints should not be allowed in case a blinded path is specified.

ziggie1984 commented on itest/list_on_test.go:
```diff
@@ -327,6 +327,10 @@ var allTestCases = []*lntest.TestCase{
 		TestFunc: testQueryRoutes,
 	},
+	{
+		Name:     "query routes routehints",
```
Using a itest here is overkill, it is already tested in: https://github.com/lightningnetwork/lnd/blob/3707b1fb7032d98dc2a8e881d3ae83e61fbc35e3/lnrpc/routerrpc/router_backend_test.go#L78

ziggie1984 replied:
Therefore commit can be dropped.

ziggie1984 commented on itest/list_on_test.go:
```diff
@@ -411,6 +415,10 @@ var allTestCases = []*lntest.TestCase{
 		TestFunc: testSendPaymentKeysendMPPFail,
 	},
+	{
+		Name:     "send payment routehints keysend",
```
also this one, we don't need this, you change focused on the `lncli` only, no need to test the already working `lnd` side. Can be removed.

ziggie1984 commented on docs/release-notes/release-notes-0.20.0.md:
```diff
@@ -28,6 +28,10 @@
 
 ## lncli Additions
 
+* [`lncli sendpayment --keysend` and `lncli queryroutes` now support the
```
needs to be changed because we will allow also `amp` payments not only keysend ...

ziggie1984 reviewed (changes requested):
Cool that you chose to contribute to LND 🤝,

I think this change can be way smaller (without itest) because you are only touching the `lncli`. Moreover the functionality is already tested on the LND-server side.

appilon commented:
@ziggie1984 I have addressed all comments. I'm a Go dev looking to break into bitcoin development, I hope to contribute more significant changes to LND and taproot-assets in the future (love what Lightning Labs is doing). Thank you for the review.

appilon commented on docs/release-notes/release-notes-0.20.0.md:
```diff
@@ -28,6 +28,10 @@
 
 ## lncli Additions
 
+* [`lncli sendpayment --keysend/--amp` and `lncli queryroutes` now support the
```
See other response

ziggie1984 reviewed (approved):
LGTM
